File: workbench_cpp/tools_graph.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
using matll = vector<vector<ll>>;

#define all(v) v.begin(), v.end()
#define min_itr(v) *min_element(v.begin(), v.end())
#define max_itr(v) *max_element(v.begin(), v.end())
#define sum(v) accumulate(v.begin(), v.end(), 0LL)
#define sort_all(v) sort(v.begin(), v.end())
#define rep(i, n) for (ll i = 0; i < ll(n); i++)
#define repi(i, a, b) for (ll i = ll(a); i < ll(b); i++)
#define for_itr(id, itr) for (auto& id : itr)
#define for_dic(key, val, dic) for (const auto& [key, val] : dic)
#define deg_to_rad(deg) (((deg) / 360) * 2 * M_PI)
#define rad_to_deg(rad) (((rad) / 2 / M_PI) * 360)
#define coutdeci cout << fixed << setprecision(15)

//隣接リスト push_backすると参照壊れるから中身だけコピペ
int nearlist(const matll& lst, matll& near) {
    for_itr(id, lst) {
        near[id[0] - 1].push_back(id[1] - 1);
        near[id[1] - 1].push_back(id[0] - 1);
    }
    return 0;
}

//幅優先探索
int bfs(const int& s, const matll& near, vector<ll>& res) {
    res[s] = 0;
    deque<ll> que;
    que.push_back(s);
    while (!que.empty()) {
        ll q;
        q = que.front();
        que.pop_front();
        for_itr(id, near[q]) {
            if (res[id] > -1) continue;
            res[id] = res[q] + 1;
            que.push_back(id);
        }
    }
    return 0;
}

//ダイクストラ法
int dijkstra(const int& s, const vector<matll>& w_near, vector<ll>& dist) {
    dist[s] = 0;
    priority_queue<vector<ll>> que;
    que.push({0, s});
    while (que.size() > 0) {
        ll d = que.top()[0], q = que.top()[1];
        que.pop();
        if (dist[q] < -d) continue;
        for_itr(nq, w_near[q]) {
            ll nxt = nq[0], tmp = -d + nq[1];
            if (dist[nxt] > tmp) {
                dist[nxt] = tmp;
                que.push({-tmp, nxt});
            }
        }
    }
    return 0;
}

// Unionfind
class Unionfind {
   private:
    vector<ll> res;
    map<ll, vector<ll>> group;

   public:
    ll N;
    vector<ll> parents;

    Unionfind(ll n0) {
        N = n0;
        parents = vector<ll>(N, -1);
    }

    ll find(ll x) {
        if (parents[x] < 0)
            return x;
        else {
            parents[x] = find(parents[x]);
            return parents[x];
        }
    }

    void unite(ll x, ll y) {
        x = find(x), y = find(y);
        if (x == y) return;
        if (parents[x] > parents[y]) {
            ll tmp = x;
            x = y, y = tmp;
        }
        parents[x] += parents[y];
        parents[y] = x;
    }

    bool same(ll x, ll y) { return (find(x) == find(y)); }

    ll roots_cnt() {
        ll cnt = 0;
        for_itr(pi, parents) if (pi < 0) cnt++;
        return cnt;
    }

    vector<ll>& roots() {
        res = vector<ll>(0);
        rep(i, N) if (parents[i] < 0) res.push_back(i);
        return res;
    }

    ll size(ll x) { return -parents[find(x)]; }

    vector<ll>& all_sizes() {
        res = vector<ll>(0);
        for_itr(pi, parents) if (pi < 0) res.push_back(-pi);
        return res;
    }

    vector<ll>& member(ll x) {
        res = vector<ll>(0);
        ll root = find(x);
        rep(i, N) if (find(i) == root) res.push_back(i);
        return res;
    }

    map<ll, vector<ll>>& all_members() {
        group = {};
        rep(i, N) if (parents[i] < 0) group[i] = vector<ll>(0);
        rep(i, N) group[find(i)].push_back(i);
        return group;
    }
};

// 強連結成分分解 rv_near: 逆向き枝, order: 帰りがけ順
class Strongly_Conected_Component {
   private:
    ll n;
    vector<ll> order, flag_dfs, flag_rdfs;
    matll near, nm_near, rv_near;

   public:
    ll cnt = 0;
    vector<ll> idx;

    Strongly_Conected_Component(ll n0, matll near0) {
        n = n0;
        near = near0;
        nm_near = near0;
        rv_near = matll(n, vector<ll>(0));
        rep(i, n) for_itr(j, near0[i]) rv_near[j].push_back(i);

        flag_dfs = vector<ll>(n, 0), order = vector<ll>(0);
        rep(i, n) if (!flag_dfs[i]) dfs(i);

        reverse(all(order));
        flag_rdfs = vector<ll>(n, 0), idx = vector<ll>(n, -1);
        for_itr(i, order) if (!flag_rdfs[i]) rdfs(i, cnt), cnt++;
    }

    void dfs(ll v) {
        flag_dfs[v] = 1;
        vector<ll> stack = {v};

        while (!stack.empty()) {
            ll now = stack.back();
            if (nm_near[now].empty()) {
                stack.pop_back();
                order.push_back(now);
            }
            while (!nm_near[now].empty()) {
                ll nxt = nm_near[now].back();
                nm_near[now].pop_back();
                if (!flag_dfs[nxt]) {
                    flag_dfs[nxt] = 1;
                    stack.push_back(nxt);
                    break;
                }
            }
        }
        return;
    }

    void rdfs(ll v, ll c) {
        idx[v] = c;
        vector<ll> stack = {v};

        while (!stack.empty()) {
            ll now = stack.back();
            stack.pop_back();
            if (flag_rdfs[now]) continue;
            flag_rdfs[now] = 1;
            for_itr(nxt, rv_near[now]) if (!flag_rdfs[nxt]) {
                idx[nxt] = c;
                stack.push_back(nxt);
            }
        }
        return;
    }
    // グラフ縮約
    matll construct() {
        matll graph(cnt, vector<ll>(0));
        rep(v, n) {
            ll v_id = idx[v];
            for_itr(w, near[v]) {
                ll w_id = idx[w];
                if (v_id != w_id) graph[v_id].push_back(w_id);
            }
        }
        return graph;
    }
};

int main() {
    ll n;
    cin >> n;

    return 0;
}

File: workbench_py/tools_graph.py
def nearlist(n0, lst0):  # 隣接リスト
    res = [set() for _ in range(n0)]
    for a, b in lst0:
        res[a - 1].add(b - 1)
        res[b - 1].add(a - 1)
    return res


def weighted_nearlist(n0, lst0):  # 重み付き隣接リスト
    res = [set() for _ in range(n0)]
    for a, b, w in lst0:
        res[a - 1].add((b - 1, w))
    return res


def bfs(s0, n0, near0):  # 幅優先探索  # キュー
    # 始点, 頂点数, 隣接リスト
    from collections import deque

    dist = [-1] * n0
    path = [-1] * n0
    flag = [0] * n0
    dist[s0], path[s0], flag[s0] = 0, 's', 1
    que = deque([s0])

    while que:
        q = que.popleft()
        for i in near0[q]:
            if flag[i]:
                continue
            # 処理を行う
            flag[i] = 1
            que.append(i)
    return


def dfs(S, N, NEAR):  # 深優先探索  # スタック
    # 始点, 頂点数, 隣接リスト
    dist = [-1] * N  # 前処理
    path = [-1] * N
    flag = [0] * N
    dist[S], path[S] = 0, 's'
    flag[S] = 1
    stack = [S]

    while stack:
        q = stack.pop()
        for i in NEAR[q]:  # 移動先の候補
            if flag[i]:  # 処理済みか否か
                continue
            # 処理を行う
            flag[i] = 1
            stack.append(i)
    return


class Recursive_dfs():  # 深優先探索(再帰)
    import sys
    sys.setrecursionlimit(10 ** 7)

    def __init__(self, S, N, NEAR):
        # 始点, 頂点数, 隣接リスト
        self.flag = [0] * N  # 前処理
        self.flag[S] = 1
        self.near = NEAR

    def recdfs(self, p):
        for i in self.near[p]:  # 移動先の候補
            if self.flag[i]:  # 処理済みか否か
                continue
            # 処理を行う
            self.flag[i] = 1
            self.recdfs(i)


def is_bipartite(S, N, NEAR):  # 二部グラフ判定
    # 始点, 頂点数, 隣接リスト
    color = [0 for i in range(N)]
    stack = [(S, 1)]
    while stack:
        q, c = stack.pop()
        for i in NEAR[q]:
            if color[i] == c:
                return False
            if color[i] == 0:
                color[i] = -c
                stack.append((i, -c))
    return True


def dijkstra(S, N, NEAR):  # ダイクストラ法:単一始点最短経路 O((n+e)*logn)
    # NEAR:隣接リスト
    from heapq import heappop, heappush
    DIST, prev = [pow(10, 10)] * N, [-1] * N
    DIST[S], prev[S] = 0, 's'

    que = [(DIST[S], S)]
    while que:
        d, q = heappop(que)
        if DIST[q] < d:
            continue
        for i, d_qi in NEAR[q]:
            tmp = d + d_qi
            if DIST[i] > tmp:
                DIST[i] = tmp
                prev[i] = q
                heappush(que, (tmp, i))
    return DIST


def warshallfloyd(N, LIST):  # ワーシャルフロイド法:全頂点対最短経路 O(n**3)
    # LIST:隣接行列
    from copy import deepcopy
    DIST = deepcopy(LIST)
    for k in range(N):
        for i in range(N):
            for j in range(N):
                DIST[i][j] = min(DIST[i][j], DIST[i][k] + DIST[k][j])
    return DIST


# トポロジカルソート:DAGに適用可, near0:有向辺隣接リスト
def topological(N, near0):
    from collections import deque

    incnt = [0] * N
    child = [set() for _ in range(N)]
    for a, b in near0:
        child[a - 1].add(b - 1)
        incnt[b - 1] += 1

    tprg = []
    que = deque([i for i, num in enumerate(incnt) if num == 0])
    while que:
        q = que.popleft()
        for i in child[q]:
            incnt[i] -= 1
            if incnt[i] == 0:
                que.append(i)
        tprg.append(q)
    return tprg


# プリム法:最小全域木
def prim(n0, near0):
    from heapq import heappush, heappop, heapify

    flag = [0] * n0
    flag[0] = 1
    que = [(c, j, 0) for j, c in near0[0]]
    heapify(que)

    ans = []
    while que:
        c_pq, q, p = heappop(que)
        if flag[q]:
            continue
        flag[q] = 1
        ans.append((p, q, c_pq))
        for r, c_qr in near0[q]:
            if 1 - flag[r]:
                heappush(que, (c_qr, r, q))
    return ans


# クラスカル法:最小全域木, UF木が必要
def kruskal(k_uf, near0):
    edge = []
    for i, ni in enumerate(near0):
        edge += [(c, i, j) for j, c in ni if i < j]
    edge = sorted(edge)

    res = 0
    for w, i, j in edge:
        if not k_uf.same(i, j):
            res += w
            k_uf.unite(i, j)
    return res


# Unionfind
class Unionfind():
    # find:グループの根, unite:グループの併合, same:同じグループか否か
    def __init__(self, n0):
        self.n = n0
        self.parents = [-1] * n0

    def find(self, x):
        if self.parents[x] < 0:
            return x
        else:
            self.parents[x] = self.find(self.parents[x])
            return self.parents[x]

    def unite(self, x, y):
        x = self.find(x)
        y = self.find(y)
        if x == y:
            return
        if self.parents[x] > self.parents[y]:
            x, y = y, x
        self.parents[x] += self.parents[y]
        self.parents[y] = x

    def same(self, x, y):
        return self.find(x) == self.find(y)

    # roots_cnt:根の数, roots:根のリスト
    def roots_cnt(self):
        return sum(x < 0 for x in self.parents)

    def roots(self):
        return [i for i, x in enumerate(self.parents) if x < 0]

    # size:特定のグループのサイズ, all_sizes:全てのグループのサイズ
    def size(self, x):
        return - self.parents[self.find(x)]

    def all_sizes(self):
        return {i: -x for i, x in enumerate(self.parents) if x < 0}

    # member:特定のグループの要素, all_members:全てのグループごとの要素
    def member(self, x):
        root = self.find(x)
        return [i for i in range(self.n) if self.find(i) == root]

    def all_members(self):
        group = {i: set() for i, x in enumerate(self.parents) if x < 0}
        for i in range(self.n):
            group[self.find(i)].add(i)
        return group


# 強連結成分分解
class strongly_conected_component():

    # near0:有向辺隣接リスト, rv_near: 逆向き枝, order: 帰りがけ順
    def __init__(self, n0, near0):
        self.n = n0
        self.near = [ni[:] for ni in near0]
        self.nm_near = [iter(ni) for ni in near0]
        self.rv_near = [[] for _ in range(self.n)]
        for i in range(self.n):
            for j in near0[i]:
                self.rv_near[j].append(i)

        self.flag_dfs = [0] * self.n
        self.order = []
        for i in range(self.n):
            if not self.flag_dfs[i]:
                self.dfs(i)

        self.cnt = 0
        self.flag_rdfs = [0] * self.n
        self.idx = [-1] * self.n
        for i in self.order[::-1]:
            if not self.flag_rdfs[i]:
                self.rdfs(i)
                self.cnt += 1
        return

    def dfs(self, v):
        self.flag_dfs[v] = 1
        stack = [v]

        while stack:
            now = stack[-1]
            for nxt in self.nm_near[now]:
                if not self.flag_dfs[nxt]:
                    self.flag_dfs[nxt] = 1
                    stack.append(nxt)
                    break
            else:
                stack.pop()
                self.order.append(now)
        return

    def rdfs(self, v):
        self.idx[v] = self.cnt
        stack = [v]

        while stack:
            now = stack.pop()
            if self.flag_rdfs[now]:
                continue
            self.flag_rdfs[now] = 1
            for nxt in self.rv_near[now]:
                if not self.flag_rdfs[nxt]:
                    self.idx[nxt] = self.cnt
                    stack.append(nxt)
        return

    # グラフ縮約
    def construct(self):
        graph = [set() for _ in range(self.cnt)]
        for v in range(self.n):
            v_id = self.idx[v]
            for w in self.near[v]:
                w_id = self.idx[w]
                if v_id != w_id:
                    graph[v_id].add(w_id)
        return graph
